controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -190,222 +188,149 @@ public void cancelChange(ApplicationId applicationId, boolean keepApplicationCha
 
     //--- End of methods which triggers deployment jobs ----------------------------
 
-    /** Find the next step to trigger if any, and triggers it */
+    /**
+     * Finds the next step to trigger for the given application, if any, and triggers it
+     */
     private void triggerReadyJobs(LockedApplication application) {
-        List<JobType> jobs =  order.jobsFrom(application.deploymentSpec());
-
-        // Should the first step be triggered?
-        if ( ! jobs.isEmpty() && jobs.get(0).equals(JobType.systemTest) ) {
-            JobStatus systemTestStatus = application.deploymentJobs().jobStatus().get(JobType.systemTest);
-            if (application.change().platform().isPresent()) {
-                Version target = application.change().platform().get();
-                if (systemTestStatus == null
-                    || ! systemTestStatus.lastTriggered().isPresent()
-                    || ! systemTestStatus.isSuccess()
-                    || ! systemTestStatus.lastTriggered().get().version().equals(target)
-                    || systemTestStatus.isHanging(jobTimeoutLimit())) {
-                    application = trigger(new Triggering(application, JobType.systemTest, false, "Upgrade to " + target), Collections.emptySet(), false);
-                    applications().store(application);
+        List<Triggering> triggerings = new ArrayList<>();
+        Change change = application.change();
+
+        // Urgh. Empty spec means unknown spec. Should we write it at component completion?
+        List<DeploymentSpec.Step> steps = application.deploymentSpec().steps();
+        if (steps.isEmpty()) steps = Collections.singletonList(new DeploymentSpec.DeclaredZone(Environment.test));
+
+        Optional<Instant> completedAt = Optional.of(clock.instant());
+        String reason = "Deploying " + change.toString();
+
+        for (DeploymentSpec.Step step : steps) {
+            LockedApplication app = application;
+            Collection<JobType> stepJobs = step.zones().stream().map(order::toJob).collect(toSet());
+            Collection<JobType> remainingJobs = stepJobs.stream().filter(job -> ! completedAt(app, job).isPresent()).collect(toList());
```

Comment:
Found this a bit confusing as type does not indicate whether order matters. I suggest using same type as the collector. `toSet` -> `Set`, `toList` -> `List`.

Reply:
Sure!